client/sniffer/src/main/java/org/elasticsearch/client/sniff/ElasticssearchHostsSniffer.java

```diff
@@ -0,0 +1,144 @@
+            JsonParser parser = jsonFactory.createParser(inputStream);
+            if (parser.nextToken() != JsonToken.START_OBJECT) {
+                throw new IOException("expected data to start with an object");
+            }
+            List<HttpHost> hosts = new ArrayList<>();
+            while (parser.nextToken() != JsonToken.END_OBJECT) {
+                if (parser.getCurrentToken() == JsonToken.START_OBJECT) {
+                    if ("nodes".equals(parser.getCurrentName())) {
+                        while (parser.nextToken() != JsonToken.END_OBJECT) {
+                            JsonToken token = parser.nextToken();
+                            assert token == JsonToken.START_OBJECT;
+                            String nodeId = parser.getCurrentName();
+                            HttpHost sniffedHost = readHost(nodeId, parser, this.scheme);
+                            if (sniffedHost != null) {
+                                logger.trace("adding node [" + nodeId + "]");
+                                hosts.add(sniffedHost);
+                            }
+                        }
+                    } else {
+                        parser.skipChildren();
+                    }
+                }
+            }
+            return hosts;
+        }
+    }
+
+    private static HttpHost readHost(String nodeId, JsonParser parser, Scheme scheme) throws IOException {
+        HttpHost httpHost = null;
+        String fieldName = null;
+        while (parser.nextToken() != JsonToken.END_OBJECT) {
+            if (parser.getCurrentToken() == JsonToken.FIELD_NAME) {
+                fieldName = parser.getCurrentName();
+            } else if (parser.getCurrentToken() == JsonToken.START_OBJECT) {
+                if ("http".equals(fieldName)) {
+                    while (parser.nextToken() != JsonToken.END_OBJECT) {
+                        if (parser.getCurrentToken() == JsonToken.VALUE_STRING && "publish_address".equals(parser.getCurrentName())) {
+                            URI boundAddressAsURI = URI.create(scheme + "://" + parser.getValueAsString());
+                            httpHost = new HttpHost(boundAddressAsURI.getHost(), boundAddressAsURI.getPort(),
+                                    boundAddressAsURI.getScheme());
+                        } else if (parser.getCurrentToken() == JsonToken.START_OBJECT) {
+                            parser.skipChildren();
+                        }
+                    }
+                } else {
+                    parser.skipChildren();
+                }
+            }
+        }
+        //http section is not present if http is not enabled on the node, ignore such nodes
+        if (httpHost == null) {
+            logger.debug("skipping node [" + nodeId + "] with http disabled");
+            return null;
+        }
+        return httpHost;
+    }
+
+    public enum Scheme {
+        HTTP("http"), HTTPS("https");
```

Comment:
not sure, I think it should not cause HostsSniffer doesn't necessarily have to make a connection etc. it can be anything. and the fact that the scheme is not part of what is returned in nodes info is very specific to es. So other impls should not need it